fpmms/profitability.py

```diff
@@ -0,0 +1,631 @@
+
+    # make sure trader_address is in the columns
+    assert "trader_address" in fpmmTrades.columns, "trader_address column not found"
+
+    # lowercase and strip creator_address
+    fpmmTrades["trader_address"] = fpmmTrades["trader_address"].str.lower().str.strip()
+
+    return fpmmTrades, tools
+
+
+def determine_market_status(trade, current_answer):
+    """Determine the market status of a trade."""
+    if current_answer is np.nan and time.time() >= trade["fpmm.openingTimestamp"]:
+        return MarketState.PENDING
+    elif current_answer == np.nan:
+        return MarketState.OPEN
+    elif trade["fpmm.isPendingArbitration"]:
+        return MarketState.ARBITRATING
+    elif time.time() < trade["fpmm.answerFinalizedTimestamp"]:
+        return MarketState.FINALIZING
+    return MarketState.CLOSED
+
+
+def analyse_trader(
+    trader_address: str, fpmmTrades: pd.DataFrame, tools: pd.DataFrame
+) -> pd.DataFrame:
+    """Analyse a trader's trades"""
+    # Filter trades and tools for the given trader
+    trades = fpmmTrades[fpmmTrades["trader_address"] == trader_address]
+    tools_usage = tools[tools["trader_address"] == trader_address]
+
+    # Prepare the DataFrame
+    trades_df = pd.DataFrame(columns=ALL_TRADES_STATS_DF_COLS)
+    if trades.empty:
+        return trades_df
+
+    # Fetch user's conditional tokens gc graph
+    try:
+        user_json = _query_conditional_tokens_gc_subgraph(trader_address)
+    except Exception as e:
+        print(f"Error fetching user data: {e}")
+        return trades_df
+
+    # Iterate over the trades
+    for i, trade in tqdm(trades.iterrows(), total=len(trades), desc="Analysing trades"):
+        try:
+            # Parsing and computing shared values
+            creation_timestamp_utc = datetime.datetime.fromtimestamp(
+                trade["creationTimestamp"], tz=datetime.timezone.utc
+            )
+            collateral_amount = wei_to_unit(float(trade["collateralAmount"]))
+            fee_amount = wei_to_unit(float(trade["feeAmount"]))
+            outcome_tokens_traded = wei_to_unit(float(trade["outcomeTokensTraded"]))
+            earnings, winner_trade = (0, False)
+            redemption = _is_redeemed(user_json, trade)
+            current_answer = trade["fpmm.currentAnswer"]
+
+            # Determine market status
+            market_status = determine_market_status(trade, current_answer)
+
+            # Skip non-closed markets
+            if market_status != MarketState.CLOSED:
+                print(
+                    f"Skipping trade {i} because market is not closed. Market Status: {market_status}"
+                )
+                continue
+            current_answer = convert_hex_to_int(current_answer)
+
+            # Compute invalidity
+            is_invalid = current_answer == INVALID_ANSWER
+
+            # Compute earnings and winner trade status
+            if is_invalid:
+                earnings = collateral_amount
+                winner_trade = False
```

Comment:
Is this case for when a market is invalidated? If yes, should we include this in the calculation as we currently do here?

https://github.com/valory-xyz/scripts/blob/2c6df71724a931b64acf9939f3b7a76c9f05090d/fpmms/profitability.py#L506-L511

cc @jmoreira-valory

Reply:
Summary after a discussion with @Adamantios:

We need to decide if (1) invalid markets are counted on the overall evaluation of the tools [as it is in the current code], or (2) if to skip them at this point.

The reasoning for (2) is that it is "unfair" for the tools, since they don't know beforehand if the market is going to be invalid or not; they simply generate a prediction and later, the market becomes invalidated.  

On the other hand, the reasoning for (1) (i.e., keep it as it is) could be because the tools should have been "smarter" to identify that the market was going to be invalidated...

For the moment, we agreed to keep the code as it is.